src/soc/ibm/power9/homer.c

```diff
@@ -3294,43 +3266,76 @@ uint64_t build_homer_image(void *homer_bar)
 			[4]	CPPM_CPMMR_RESERVED_2_9	= 1
 		*/
 		/* SCOM2 - OR, 0x200F0108 */
-		write_scom_for_chiplet(EC00_CHIPLET_ID + i, 0x200F0108,
-		                       PPC_BIT(3) | PPC_BIT(4));
+		write_rscom_for_chiplet(chip, EC00_CHIPLET_ID + i, 0x200F0108,
+					PPC_BIT(3) | PPC_BIT(4));
 	}
+}
 
-	/* 15.2 set HOMER BAR */
-	report_istep(15,2);
-	write_scom(0x05012B00, (uint64_t)homer);
-	write_scom(0x05012B04, (4 * MiB - 1) & ~((uint64_t)MiB - 1));
-	write_scom(0x05012B02, (uint64_t)homer + 8 * 4 * MiB);		// FIXME
-	write_scom(0x05012B06, (8 * MiB - 1) & ~((uint64_t)MiB - 1));
+/* 15.2 set HOMER BAR */
+static void istep_15_2(uint8_t chip, struct homer_st *homer)
+{
+	write_rscom(chip, 0x05012B00, (uint64_t)homer);
+	write_rscom(chip, 0x05012B04, (4 * MiB - 1) & ~((uint64_t)MiB - 1));
 
-	/* 15.3 establish EX chiplet */
-	report_istep(15,3);
-	/* Multicast groups for cores were assigned in get_available_cores() */
-	for (int i = 0; i < MAX_QUADS_PER_CHIP; i++) {
-		if (IS_EQ_FUNCTIONAL(i, cores) &&
-		    (read_scom_for_chiplet(EP00_CHIPLET_ID + i, 0xF0001) & PPC_BITMASK(3,5))
-		    == PPC_BITMASK(3,5))
-			scom_and_or_for_chiplet(EP00_CHIPLET_ID + i, 0xF0001,
-			                        ~(PPC_BITMASK(3,5) | PPC_BITMASK(16,23)),
+	write_rscom(chip, 0x05012B02, (uint64_t)homer + (8 - chip) * 4 * MiB);		// FIXME
```

Comment:
This address (OCC common area or whatever its official name is) should be passed by the caller.

I think we should rework calculation for HOMER base. I hardcoded it to the way Hostboot sets it so I can reuse its image but that is no longer necessary. We can reserve only as much as is needed (8M + 4M * number of chips), instead of always reserving 40M (IIRC). This can be done now or as separate change later, whichever is easier.

Reply:
I don't think it will complicate rebasing, so done it here.